hermes/src/store/storage/local_storage.rs

```diff
@@ -0,0 +1,163 @@
+use {
+    super::{
+        AccumulatorState,
+        MessageState,
+        MessageStateFilter,
+        MessageStateKey,
+        RequestTime,
+        Storage,
+        StorageInstance,
+    },
+    crate::store::types::Slot,
+    anyhow::{
+        anyhow,
+        Result,
+    },
+    async_trait::async_trait,
+    dashmap::DashMap,
+    moka::sync::Cache,
+    pyth_oracle::MessageType,
+    pyth_sdk::PriceIdentifier,
+    std::{
+        collections::VecDeque,
+        sync::Arc,
+    },
+    strum::IntoEnumIterator,
+};
+
+#[derive(Clone)]
+pub struct LocalStorage {
+    message_cache:     Arc<DashMap<MessageStateKey, VecDeque<MessageState>>>,
+    accumulator_cache: Cache<Slot, AccumulatorState>,
+    cache_size:        u64,
+}
+
+impl LocalStorage {
+    pub fn new_instance(cache_size: u64) -> StorageInstance {
+        Box::new(Self {
+            message_cache: Arc::new(DashMap::new()),
+            accumulator_cache: Cache::builder()
+                .max_capacity(cache_size)
+                .time_to_live(std::time::Duration::from_secs(60 * 60 * 24))
+                .build(),
+            cache_size,
+        })
+    }
+
+    fn retrieve_message_state(
+        &self,
+        key: MessageStateKey,
+        request_time: RequestTime,
+    ) -> Option<MessageState> {
+        match self.message_cache.get(&key) {
+            Some(key_cache) => {
+                match request_time {
+                    RequestTime::Latest => key_cache.back().cloned(),
+                    RequestTime::FirstAfter(time) => {
+                        // If the requested time is before the first element in the vector, we are
+                        // not sure that the first element is the closest one.
+                        if let Some(oldest_record) = key_cache.front() {
+                            if time < oldest_record.time().publish_time {
+                                return None;
+                            }
+                        }
+
+                        // Binary search returns Ok(idx) if the element is found at index idx or Err(idx) if it's not
+                        // found which idx is the index where the element should be inserted to keep the vector sorted.
+                        // Getting idx within any of the match arms will give us the index of the element that is
+                        // closest after or equal to the requested time.
+                        let idx = match key_cache
+                            .binary_search_by_key(&time, |record| record.time().publish_time)
```

Comment:
note that we need to secondarily sort by slot here. If the records are stored in ascending order of (publish_time, slot), i think we need to search for (publish_time, 0)

Reply:
good catch. It is sorted by slot as well but apparently binary search doesn't guarantee that it gives you the first. I'll fix it.